orttraining/orttraining/python/orttraining_pybind_state.cc

```diff
@@ -357,16 +362,106 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
       .def("reserve", [](std::vector<OrtValue>* v, const size_t len) { v->reserve(len); })
       .def("shrink_to_fit", [](std::vector<OrtValue>* v) { v->shrink_to_fit(); })
       .def("__len__", [](const std::vector<OrtValue>& v) { return v.size(); })
-      .def("__iter__", [](const std::vector<OrtValue>& v) {
-        return py::make_iterator(v.cbegin(), v.cend());
-      },
-           py::keep_alive<0, 1>())
+      .def(
+          "__iter__", [](const std::vector<OrtValue>& v) {
+            return py::make_iterator(v.cbegin(), v.cend());
+          },
+          py::keep_alive<0, 1>())
       .def("__getitem__", [](const std::vector<OrtValue>& v, const size_t idx) {
         return v.at(idx);
       })
       .def("dlpack_at", [](std::vector<OrtValue>* v, const size_t idx) {
         return py::reinterpret_steal<py::object>(ToDlpack(v->at(idx)));
-      });
+      })
+      .def("to_dlpack", [](const std::vector<OrtValue>& v, py::object to_tensor) -> py::list {
+
+        if (v.size() == 0)
+          return py::list();
+
+        py::list res;
+
+        PyObject* capsule;
+        PyObject* handle = to_tensor.ptr();
+        PyObject* obj;
+
+        py::gil_scoped_acquire acquire;
+
+        const Tensor& tensor = v[0].Get<Tensor>();
+
+        if (tensor.Location().device.Type() == OrtDevice::CPU) {
+          OrtDLManagedTensor ort_dlmanaged_tensor;
+          capsule = PyCapsule_New(&ort_dlmanaged_tensor.tensor, "dltensor", NULL);
+          if (capsule == NULL)
+            throw std::runtime_error("Empty capsule returned.");
+
+          for (auto it : v) {
+            // The same capsule is reused but FromDLPack rename the capsule into used_dltensor.
+            PyCapsule_SetName(capsule, "dltensor");
+            dlpack::OrtValueToDlpack(it, (void*)&ort_dlmanaged_tensor);
+    #if (PY_MAJOR_VERSION >= 3 && PY_MINOR_VERSION >= 9)
+            obj = PyObject_CallOneArg(handle, capsule);
+    #else
+            obj = PyObject_CallFunctionObjArgs(handle, capsule, NULL);
+    #endif
+            if (obj == NULL)
+              throw std::runtime_error("Empty tensor returned.");
+            res.append(py::handle(obj));
```

Comment:
do we need decrease obj ref cnt? assuming res.append() will increase the obj's ref_cnt by 1.

Reply:
We need obj to persist after the execution of this function. obj is a torch tensor. It will be deleted when it is no longer needed later.

Reply:
oh, I mean. after "obj = PyObject_CallOneArg" runs, obj as a PyObject* will have ref_cnt = 1. And the ref cnt will be decreased until we explicitly call Py_DECREF. 

On the other hand, I am not that sure, but have following guess: if we append obj into res, Py_REFCNT(obj) will be 2?  So after the function returns res, obj still have REF_CNT = 1. You can add add Py_REFCNT in multiple places to see the reference count.

Reply:
You are right. PyList_Append increases the reference count.

Reply:
However function PyObject_CallOneArg does not return a new reference, so I'm not sure about the count anymore. I used it to be faster. I'll remove it.